File: 3743-reschedule-meetings-for-maximum-free-time-i/3743-reschedule-meetings-for-maximum-free-time-i.py
class Solution:
    def maxFreeTime(self, eventTime: int, k: int, startTime: List[int], endTime: List[int]) -> int:
        # Approach 1: Sliding Window

        n = len(startTime)

        # gaps = [0] * (n+1)

        gaps = []
        # gaps[0] = startTime[0] - 0
        gaps.append(startTime[0] - 0)

        for i in range(n-1):
            # gaps[i] = startTime[i] - endTime[i-1]
            gaps.append(startTime[i+1] - endTime[i])
        
        # gaps[n] = eventTime - endTime[n-1]
        gaps.append(eventTime - endTime[n-1])

        window_size = k + 1

        if len(gaps) <= window_size:
            return sum(gaps)
        
        current_window_gap_sum = sum(gaps[:window_size])
        max_free_time = current_window_gap_sum

        for i in range(1, len(gaps) - window_size + 1):
            current_window_gap_sum = current_window_gap_sum - gaps[i-1] + gaps[i + window_size - 1]

            max_free_time = max(current_window_gap_sum, max_free_time)
        
        return max_free_time

File: 3743-reschedule-meetings-for-maximum-free-time-i/3743-reschedule-meetings-for-maximum-free-time-i.go
func maxFreeTime(eventTime int, k int, startTime []int, endTime []int) int {
    n := len(startTime)

    gaps := make([]int, 0, n+1)

    gaps = append(gaps, startTime[0] - 0)

    for i := 0; i < n-1; i++ {
        gaps = append(gaps, startTime[i + 1] - endTime[i])
    }

    gaps = append(gaps, eventTime - endTime[n-1])

    windowSize := k + 1 // for k moves, the total no.of gaps contributed = +1

    if len(gaps) <= windowSize {
        return Sum(gaps)
    }

    currentWindowSum := Sum(gaps[:windowSize])
    maxFreeTime := currentWindowSum

    for i := 1; i <= len(gaps) - windowSize; i++ {
        currentWindowSum = currentWindowSum  - gaps[i-1] + gaps[i + windowSize - 1]

        maxFreeTime = max(maxFreeTime, currentWindowSum)
    }

    return maxFreeTime

}

func Sum (nums []int) int {
    totalSum := 0

    for _, num := range nums {
        totalSum += num
    }

    return totalSum
}